etor/src/main/java/gov/hhs/cdc/trustedintermediary/etor/ruleengine/transformation/custom/MapLocalObservationCodes.java

```diff
@@ -23,52 +23,68 @@
         var observations = HapiHelper.resourcesInBundle(bundle, Observation.class);
 
-        for (Observation obv : observations.toList()) {
-            var codingList = obv.getCode().getCoding();
+        observations
+                .filter(this::hasValidCoding)
+                .forEach(
+                        observation ->
+                                processCoding(observation, codingMap, msh41Identifier, messageId));
+    }
 
-            if (codingList.size() != 1) {
-                continue;
-            }
+    private boolean hasValidCoding(Observation observation) {
+        var codingList = observation.getCode().getCoding();
+        return codingList.size() == 1 && isLocalCode(codingList.get(0));
+    }
 
-            var coding = codingList.get(0);
-            if (!HapiHelper.hasDefinedCoding(
-                    coding, HapiHelper.EXTENSION_ALT_CODING, HapiHelper.LOCAL_CODE)) {
-                continue;
-            }
+    private boolean isLocalCode(Coding coding) {
+        return HapiHelper.hasDefinedCoding(
+                coding, HapiHelper.EXTENSION_ALT_CODING, HapiHelper.LOCAL_CODE);
+    }
 
-            var identifier = codingMap.get(coding.getCode());
-            if (identifier == null) {
-                logUnmappedLocalCode(bundle, coding);
-                continue;
-            }
+    private String extractMsh41Identifier(Bundle bundle) {
+        var msh41Identifier = HapiHelper.getMSH4_1Identifier(bundle);
+        return msh41Identifier != null ? msh41Identifier.getValue() : null;
+    }
 
-            var mappedCoding = getMappedCoding(identifier);
+    private void processCoding(
+            Observation observation,
+            Map<String, IdentifierCode> codingMap,
+            String msh41Identifier,
+            String messageId) {
+        var originalCoding = observation.getCode().getCoding().get(0);
+        IdentifierCode identifier = codingMap.get(originalCoding.getCode());
+
+        if (identifier == null) {
+            logUnmappedLocalCode(originalCoding, msh41Identifier, messageId);
+            return;
+        }
 
-            // Add the mapped code as the first in the list, ahead of the existing alternate code
-            codingList.add(0, mappedCoding);
+        var mappedCoding = getMappedCoding(identifier);
+        observation.getCode().getCoding().add(0, mappedCoding);
+    }
+
+    private String validateField(String field, String fieldName) {
+        if (field == null || field.isBlank()) {
+            throw new IllegalArgumentException("missing or empty required field " + fieldName);
         }
+        return field;
     }
 
-    private void logUnmappedLocalCode(Bundle bundle, Coding coding) {
-        var msh41Identifier = HapiHelper.getMSH4_1Identifier(bundle);
-        var msh41Value = msh41Identifier != null ? msh41Identifier.getValue() : null;
+    private void logUnmappedLocalCode(Coding coding, String msh41Identifier, String messageId) {
```

Comment:
Looks like this is only used once, it has a single line and the content is pretty straightforward. Does it need to extracted into its own function?

Reply:
Good suggestion. I'll inline the extracting method and keep the logging one as is to keep the logging logic in one area.

Reply:
What I meant is that we probably don't need `logUnmappedLocalCode` as it's adding more abstraction without the benefits. You could directly call `logger.logWarning` without putting it into its own function

Reply:
Made the logging inline. Thanks for the suggestion, I think it makes sense.